### ISSUE TITLE
Remove Branches after Merge?

### ISSUE THREAD

**Issue by dallashall** [dallashall/react-redux-generator#13]
[This link](https://www.quora.com/Should-we-remove-a-remote-Git-branch-after-merging-to-master) has some good points that vibe more with @niartenyaw's point of view. I'll let you decide.

**Comment by dobrynin** [comment]
Removing branches after merge jibes well with me.

**Comment by niartenyaw** [comment]
Yeah, after a merge with master, I don't think any more work should be done on that branch. Another one should be checked out if changes need to be made. I'm a little confused about what kind of commit history is kept though if we delete branches. Does master only track the merged commit of all changes for that branch @dallashall? If so, we may want to think about rebasing branches instead. And make sure the branch commits are clean. [Really good article about rebasing](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) and some comparison to merging at the bottom. It seems rebase is only a problem if there is a force overwrite of commits (though there are fixes). Not sure how common that is. But locally rebasing can allow us to keep very clean branch histories so the master history will be much cleaner.

**Comment by dallashall** [comment]
Good article! I also liked [this one](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request) from edX's contribution guidelines.

I think that rebasing locally, if anything has updated on the master since checking out a new branch, makes a lot of sense.

It _is_ safe to delete a branch after merging, so if everyone here agrees, we will delete branches after a successful merge.

Once we have an alpha version of the code running, we might consider a branch system that allows for 'Production' or 'Stable' to be updated in larger chunks, and then have 'Testing', 'Beta', or 'Development' branches for the bulk of the work. These would be branches that are never deleted.

TL;DR:
- If you have finished working on a branch, and no changes have been made to the master branch, no rebase will be necessary.
- If you have finished working on a branch, and changes have been made to the master branch, prefer to rebase locally, handling any conflicts before submitting a PR.
- Once successfully merged, a branch can be deleted.

**Comment by niartenyaw** [comment]
Also a great one! I totally agree with the TL;DR above.

**Comment by dallashall** [comment]
Just to be 100% clear...

We still merge PRs here on GitHub. Rebasing should be done locally before pushing. That should give us the most complete history while keeping everything clean.
